File: RealtimeObjectDetectorFlutter/realtime_object_detector_flutter/lib/model_runner.dart
import 'dart:math';
import 'dart:typed_data';

import 'package:flutter/cupertino.dart';
import 'package:realtime_object_detector_flutter/detection.dart';
import 'package:tflite_flutter/tflite_flutter.dart';

class ModelRunner {
  late Interpreter _interpreter;
  late List<String> _labels;

  void loadModelFromBytes(Uint8List modelBytes, List<String> labels) {
    _interpreter = Interpreter.fromBuffer(modelBytes);
    _labels = labels;
    _logModelInfo();
  }

  void _logModelInfo() {
    final inputTensor = _interpreter.getInputTensor(0);

    debugPrint("INPUT SHAPE: ${inputTensor.shape}");
    debugPrint('INPUT TYPE: ${inputTensor.type}');

    final outputCount = _interpreter.getOutputTensors().length;

    debugPrint("OUTPUT COUNT: $outputCount");

    for (int i = 0; i < outputCount; i++) {
      final tensor = _interpreter.getOutputTensor(i);

      debugPrint(
        'OUTPUT $i -> '
        'shape=${tensor.shape}'
        'type=${tensor.type}',
      );
    }
  }

  List<Detection> detect(Map<String, dynamic> frameData) {
    final letterbox = _preprocess(frameData);
    final input = letterbox.input;
    final boxes = List.generate(
      1,
      (_) => List.generate(25, (_) => List.filled(4, 0.0)),
    );
    final classes = List.generate(1, (_) => List.filled(25, 0.0));
    final scores = List.generate(1, (_) => List.filled(25, 0.0));
    final count = List.filled(1, 0.0);

    final outputs = {0: boxes, 1: classes, 2: scores, 3: count};
    debugPrint("OUTPUTS -> ${outputs[0]}");
    _interpreter.runForMultipleInputs([input], outputs);

    return _parseDetections(
      boxes,
      classes,
      scores,
      count[0].toInt(),
      letterbox,
    );
  }

  List<Detection> _parseDetections(
    List<List<List<double>>> boxes,
    List<List<double>> classes,
    List<List<double>> scores,
    int detectionCount,
    LetterboxInfo letterbox,
  ) {
    final detections = <Detection>[];
    const double inputSize = 320.0;

    debugPrint('DETECTION COUNT RAW: $detectionCount');

    for (int i = 0; i < detectionCount; i++) {
      final score = scores[0][i];

      if (score < 0.5) continue;

      final classIndex = classes[0][i].toInt() + 1;

      final label = (classIndex >= 0 && classIndex < _labels.length)
          ? _labels[classIndex]
          : 'Unknown';

      final box = boxes[0][i];

      final ymin = box[0].clamp(0.0, 1.0) * inputSize;
      final xmin = box[1].clamp(0.0, 1.0) * inputSize;
      final ymax = box[2].clamp(0.0, 1.0) * inputSize;
      final xmax = box[3].clamp(0.0, 1.0) * inputSize;

      final left = (xmin - letterbox.padX) / letterbox.scale;
      final top = (ymin - letterbox.padY) / letterbox.scale;
      final right = (xmax - letterbox.padX) / letterbox.scale;
      final bottom = (ymax - letterbox.padY) / letterbox.scale;

      print(
        'BOX label=$label score=$score '
        'raw=${boxes[0][i]} '
        'restored=[$left,$top,$right,$bottom]',
      );

      detections.add(
        Detection(
          label: label,
          confidence: score,
          left: left,
          top: top,
          right: right,
          bottom: bottom,
        ),
      );
    }
    return detections;
  }

  LetterboxInfo _preprocess(Map<String, dynamic> frameData) {
    final int width = frameData['width'];
    final int height = frameData['height'];

    final List planes = frameData['planes'];
    final String format = frameData['format'];

    if (format == 'yuv420') {
      // Android YUV420
      return _yuv420ToLetterboxTensor(width, height, planes);
    }

    if (format == "bgra888") {
      // Ios BGRA8888
      return _bgra8888ToLetterboxTensor(width, height, planes);
    }

    throw Exception('Unsupported image format: $format');
  }

  LetterboxInfo _yuv420ToLetterboxTensor(int width, int height, List planes) {
    final yPlane = planes[0];
    final uPlane = planes[1];
    final vPlane = planes[2];

    final Uint8List yBytes = yPlane['bytes'] as Uint8List;
    final Uint8List uBytes = uPlane['bytes'] as Uint8List;
    final Uint8List vBytes = vPlane['bytes'] as Uint8List;

    final int yRowStride = yPlane['bytesPerRow'] as int;
    final int uvRowStride = uPlane['bytesPerRow'] as int;
    final int uvPixelStride = uPlane['bytesPerPixel'] as int;

    const int inputSize = 320;

    final double scale = min(inputSize / width, inputSize / height);

    final int scaledWidth = (width * scale).round();
    final int scaledHeight = (height * scale).round();

    final double padX = (inputSize - scaledWidth) / 2.0;
    final double padY = (inputSize - scaledHeight) / 2.0;

    final rgbBytes = Uint8List(inputSize * inputSize * 3);

    int outputIndex = 0;

    for (int y = 0; y < inputSize; y++) {
      for (int x = 0; x < inputSize; x++) {
        final double sourceXInScaled = x - padX;
        final double sourceYInScaled = y - padY;

        if (sourceXInScaled < 0 ||
            sourceYInScaled < 0 ||
            sourceXInScaled >= scaledWidth ||
            sourceYInScaled >= scaledHeight) {
          // Padding area: blackpixel
          rgbBytes[outputIndex++] = 0;
          rgbBytes[outputIndex++] = 0;
          rgbBytes[outputIndex++] = 0;
          continue;
        }

        final int srcX = (sourceXInScaled / scale).toInt();
        final int srcY = (sourceYInScaled / scale).toInt();

        final int safeX = srcX.clamp(0, width - 1);
        final int safeY = srcY.clamp(0, height - 1);

        final int yIndex = safeY * yRowStride + safeX;

        final int uvX = safeX ~/ 2;
        final int uvY = safeY ~/ 2;

        final int uvIndex = uvY * uvRowStride + uvX * uvPixelStride;

        final int yValue = yBytes[yIndex];
        final int uValue = uBytes[uvIndex];
        final int vValue = vBytes[uvIndex];

        final double yf = yValue.toDouble();
        final double uf = uValue.toDouble() - 128.0;
        final double vf = vValue.toDouble() - 128.0;

        int r = (yf + 1.402 * vf).round();
        int g = (yf - 0.344136 * uf - 0.714136 * vf).round();
        int b = (yf + 1.772 * uf).round();

        r = r.clamp(0, 255);
        g = g.clamp(0, 255);
        b = b.clamp(0, 255);

        rgbBytes[outputIndex++] = r;
        rgbBytes[outputIndex++] = g;
        rgbBytes[outputIndex++] = b;
      }
    }
    return LetterboxInfo(input: rgbBytes, scale: scale, padX: padX, padY: padY);
  }

  LetterboxInfo _bgra8888ToLetterboxTensor(int width, int height, List planes) {
    final plane = planes[0];
    final Uint8List bytes = plane['bytes'] as Uint8List;
    final int bytesPerRow = plane['bytesPerRow'] as int;

    const inputSize = 320;

    final double scale = min(inputSize / width, inputSize / height);

    final int scaledWidth = (width * scale).round();
    final int scaledHeight = (height * scale).round();

    final double padX = (inputSize - scaledWidth) / 2.0;
    final double padY = (inputSize - scaledHeight) / 2.0;

    final Uint8List rgbBytes = Uint8List(inputSize * inputSize * 3);

    int pixelIndex = 0;

    for (int y = 0; y < inputSize; y++) {
      for (int x = 0; x < inputSize; x++) {
        final double sourceXInScaled = x - padX;
        final double sourceYInScaled = x - padY;

        if (sourceXInScaled < 0 ||
            sourceYInScaled < 0 ||
            sourceXInScaled >= scaledWidth ||
            sourceYInScaled >= scaledHeight) {
          rgbBytes[pixelIndex++] = 0;
          rgbBytes[pixelIndex++] = 0;
          rgbBytes[pixelIndex++] = 0;
          continue;
        }

        final int srcX = (sourceXInScaled / scale).toInt();
        final int srcY = (sourceYInScaled / scale).toInt();

        final int safeX = srcX.clamp(0, width - 1);
        final int safeY = srcY.clamp(0, height - 1);

        final int byteIndex = (safeY * bytesPerRow + safeX) * 4;

        final int b = bytes[byteIndex];
        final int g = bytes[byteIndex + 1];
        final int r = bytes[byteIndex + 2];

        rgbBytes[pixelIndex++] = r;
        rgbBytes[pixelIndex++] = g;
        rgbBytes[pixelIndex++] = b;
      }
    }
    return LetterboxInfo(input: rgbBytes, scale: scale, padX: padX, padY: padY);
  }
}

class LetterboxInfo {
  final Uint8List input;
  final double scale;
  final double padX;
  final double padY;

  LetterboxInfo({
    required this.input,
    required this.scale,
    required this.padX,
    required this.padY,
  });
}

File: RealtimeObjectDetectorIOS/RealtimeObjectDetectorIOS/ModelRunner.swift
//
//  ModelRunner.swift
//  RealtimeObjectDetectorIOS
//
//  Created by Emmanuel Emmanuel on 18/05/2026.
//
import TensorFlowLite
import Foundation
import CoreVideo
import CoreGraphics

struct LetterboxInfo {
    let inputData: Data
    let scale: CGFloat
    let padX: CGFloat
    let padY: CGFloat
    let originalWidth: Int
    let originalHeight: Int
}

final class ModelRunner {
    private var interpreter: Interpreter?
    private var labels: [String] = []
    
    init() {
        loadModel()
        loadLabels()
    }
    
    private func loadModel() {
        guard let modelPath = Bundle.main.path(forResource: "efficientdet_lite0", ofType: "tflite") else {
            print("Model file not found!")
            return
        }
        
        do {
            interpreter = try Interpreter(modelPath: modelPath)
            try interpreter?.allocateTensors()
            print("Model loaded successfully")
            logModelInfo()
        } catch {
            print("Failed to load model: \(error)")
        }
    }
    
    private func loadLabels() {
        guard let path = Bundle.main.path(forResource: "labels", ofType: "txt") else {
            print("Labels file not found")
            return
        }
        
        do {
            let text = try String(contentsOfFile: path, encoding: .utf8)
            labels = text
                .components(separatedBy: .newlines)
                .map { $0.trimmingCharacters(in: .whitespacesAndNewlines)}
                .filter { !$0.isEmpty }
            print("Labels loaded: \(labels.count)")
        } catch {
            print("Faile to load labels: \(error)")
        }
    }
    
    private func logModelInfo() {
        guard let interpreter else { return }
        do {
            let inputTensor = try interpreter.input(at: 0)
            
            print("Input tensor shape: \(inputTensor.shape)")
            print("Input tensor type: \(inputTensor.dataType)")
            
            let outputCount = interpreter.outputTensorCount
            print("Output tensor count: \(outputCount)")
            
            for i in 0..<outputCount {
                let outputTensor = try interpreter.output(at: i)
                print("Output \(i) shape: \(outputTensor.shape), type: \(outputTensor.dataType)")
                
            }
        } catch {
            print("Failed to log model info: \(error)")
        }
    }
    
    func detect(pixelBuffer: CVPixelBuffer) -> [Detection] {
        guard let interpreter else { return [] }
        
        guard let letterbox = pixelBufferToLetterboxInput(pixelBuffer) else {
            return []
        }
        
        do {
            try interpreter.copy(letterbox.inputData, toInputAt: 0)
            try interpreter.invoke()
            
            let boxesTensor = try interpreter.output(at: 0)
            let classesTensor = try interpreter.output(at: 1)
            let scoresTensor = try interpreter.output(at: 2)
            let countTensor = try interpreter.output(at: 3)
            
            let boxes = boxesTensor.data.toFloatArray()
            let classes = classesTensor.data.toFloatArray()
            let scores = scoresTensor.data.toFloatArray()
            let count = Int(countTensor.data.toFloatArray().first ?? 0)
            
            return parseDetections(
                boxes: boxes,
                classes: classes,
                scores: scores,
                count: count,
                letterbox: letterbox
            )
        } catch {
            print("Detection failed: \(error)")
            return []
        }
    }
    
    private func parseDetections(
    boxes: [Float],
    classes: [Float],
    scores: [Float],
    count: Int,
    letterbox: LetterboxInfo
    
    ) -> [Detection]{
     
        var detections: [Detection] = []
        
        let inputSize: CGFloat = 320.0
        let maxDetections = min(count, 25)
        
        for i in 0..<maxDetections {
            let score = scores[i]
            if score < 0.3 { continue }
            let classIndex = Int(classes[i]) + 1
            let label: String
            if classIndex >= 0 && classIndex < labels.count {
                label = labels[classIndex]
            }else {
                label = "Unknown"
            }
            let boxOffset = i * 4
            
            let ymin = CGFloat(boxes[boxOffset+0]).clamped(to: 0...1) * inputSize
            let xmin = CGFloat(boxes[boxOffset+1]).clamped(to: 0...1) * inputSize
            let ymax = CGFloat(boxes[boxOffset+2]).clamped(to: 0...1) * inputSize
            let xmax = CGFloat(boxes[boxOffset+3]).clamped(to: 0...1) * inputSize
            
            let left = (xmin - letterbox.padX) / letterbox.scale
            let top = (ymin - letterbox.padY) / letterbox.scale
            let right = (xmax - letterbox.padX) / letterbox.scale
            let bottom = (ymax - letterbox.padY) / letterbox.scale
            
            let clippedLeft = max(0, min(left, CGFloat(letterbox.originalWidth)))
            let clippedTop = max(0, min(top, CGFloat(letterbox.originalHeight)))
            let clippedRight = max(0, min(right, CGFloat(letterbox.originalWidth)))
            let clippedBottom = max(0, min(bottom, CGFloat(letterbox.originalHeight)))
            
            let rect = CGRect(x: clippedLeft, y: clippedTop, width: clippedRight - clippedLeft, height: clippedBottom - clippedTop)
            
            detections.append(
                Detection(
                    label: label ,
                    confidence: score,
                    rect: rect
                )
            )
        }
        
        return detections
    }
    
    private func pixelBufferToLetterboxInput(_ pixelBuffer: CVPixelBuffer) -> LetterboxInfo? {
        CVPixelBufferLockBaseAddress(pixelBuffer, .readOnly)
        defer {
            CVPixelBufferUnlockBaseAddress(pixelBuffer, .readOnly)
        }

        guard let baseAddress = CVPixelBufferGetBaseAddress(pixelBuffer) else {
            return nil
        }

        let rawWidth = CVPixelBufferGetWidth(pixelBuffer)      // usually 640
        let rawHeight = CVPixelBufferGetHeight(pixelBuffer)    // usually 480
        let bytesPerRow = CVPixelBufferGetBytesPerRow(pixelBuffer)

        // After 90-degree rotation, logical image becomes portrait-ish:
        let imageWidth = rawHeight     // 480
        let imageHeight = rawWidth     // 640

        let inputSize = 320

        let scale = min(
            CGFloat(inputSize) / CGFloat(imageWidth),
            CGFloat(inputSize) / CGFloat(imageHeight)
        )

        let scaledWidth = Int((CGFloat(imageWidth) * scale).rounded())
        let scaledHeight = Int((CGFloat(imageHeight) * scale).rounded())

        let padX = CGFloat(inputSize - scaledWidth) / 2.0
        let padY = CGFloat(inputSize - scaledHeight) / 2.0

        let buffer = baseAddress.assumingMemoryBound(to: UInt8.self)

        var inputBytes = [UInt8]()
        inputBytes.reserveCapacity(inputSize * inputSize * 3)

        for y in 0..<inputSize {
            for x in 0..<inputSize {
                let sourceXInScaled = CGFloat(x) - padX
                let sourceYInScaled = CGFloat(y) - padY

                if sourceXInScaled < 0 ||
                    sourceYInScaled < 0 ||
                    sourceXInScaled >= CGFloat(scaledWidth) ||
                    sourceYInScaled >= CGFloat(scaledHeight) {

                    inputBytes.append(0)
                    inputBytes.append(0)
                    inputBytes.append(0)
                    continue
                }

                // Coordinates in rotated/upright image space
                let rotatedX = Int(sourceXInScaled / scale)
                let rotatedY = Int(sourceYInScaled / scale)

                let safeRotatedX = min(max(rotatedX, 0), imageWidth - 1)
                let safeRotatedY = min(max(rotatedY, 0), imageHeight - 1)

                // Map upright portrait coordinate back to raw landscape buffer.
                // 90° clockwise transform:
                let rawX = safeRotatedY
                let rawY = rawHeight - 1 - safeRotatedX

                let byteIndex = rawY * bytesPerRow + rawX * 4

                // BGRA memory
                let b = buffer[byteIndex]
                let g = buffer[byteIndex + 1]
                let r = buffer[byteIndex + 2]

                // Model wants RGB
                inputBytes.append(r)
                inputBytes.append(g)
                inputBytes.append(b)
            }
        }

        return LetterboxInfo(
            inputData: Data(inputBytes),
            scale: scale,
            padX: padX,
            padY: padY,
            originalWidth: imageWidth,
            originalHeight: imageHeight
        )
    }
}


extension Data {
    func toFloatArray() -> [Float] {
        return self.withUnsafeBytes { rawBuffer -> [Float] in
            // Bind the raw memory to Float
            let floatBuffer = rawBuffer.bindMemory(to: Float.self)
            return Array(floatBuffer)
        }
    }
}

extension Comparable {
    func clamped(to limits: ClosedRange<Self>) -> Self {
        min(max(self, limits.lowerBound), limits.upperBound)
    }
}
